java/src/main/native/src/row_conversion.cu

```diff
@@ -304,8 +329,620 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+  auto group = cooperative_groups::this_thread_block();
+  extern __shared__ int8_t shared_data[];
+  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_block};
+
+  __shared__ cuda::barrier<cuda::thread_scope_block> block_barrier[NUM_BLOCKS_PER_KERNEL_LOADED];
+  if (group.thread_rank() == 0) {
+    for (int i = 0; i < NUM_BLOCKS_PER_KERNEL_LOADED; ++i) {
+      init(&block_barrier[i], group.size());
+    }
+  }
+
+  group.sync();
+
+  auto const blocks_remaining =
+      std::min((uint)block_infos.size() - blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS,
+               (uint)NUM_BLOCKS_PER_KERNEL_TO_ROWS);
+
+  size_t fetch;
+  size_t subset;
+  for (subset = fetch = 0; subset < blocks_remaining; ++subset) {
+    // Fetch ahead up to stages_count subsets
+    for (; fetch < blocks_remaining && fetch < (subset + stages_count); ++fetch) {
+      auto const fetch_block = block_infos[blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS + fetch];
+      auto const num_fetch_cols = fetch_block.num_cols();
+      auto const num_fetch_rows = fetch_block.num_rows();
+      auto const num_elements_in_block = num_fetch_cols * num_fetch_rows;
+      auto const fetch_block_row_size = fetch_block.get_shared_row_size(col_offsets, col_sizes);
+      auto const starting_column_offset = col_offsets[fetch_block.start_col];
+      auto &fetch_barrier = block_barrier[fetch % NUM_BLOCKS_PER_KERNEL_LOADED];
+
+      // wait for the last use of the memory to be completed
+      if (fetch >= NUM_BLOCKS_PER_KERNEL_LOADED) {
+        fetch_barrier.arrive_and_wait();
+      }
+
+      // to do the copy we need to do n column copies followed by m element copies OR
+      // we have to do m element copies followed by r row copies. When going from column
+      // to row it is much easier to copy by elements first otherwise we would need a running
+      // total of the column sizes for our block, which isn't readily available. This makes it
+      // more appealing to copy element-wise from input data into shared matching the end layout
+      // and do row-based memcopies out.
+
+      auto const shared_buffer_base = shared[fetch % stages_count];
+      for (auto el = (int)threadIdx.x; el < num_elements_in_block; el += blockDim.x) {
+        auto const relative_col = el / num_fetch_rows;
+        auto const relative_row = el % num_fetch_rows;
+        auto const absolute_col = relative_col + fetch_block.start_col;
+        auto const absolute_row = relative_row + fetch_block.start_row;
+        auto const col_size = col_sizes[absolute_col];
+        auto const col_offset = col_offsets[absolute_col];
+        auto const relative_col_offset = col_offset - starting_column_offset;
+
+        auto const shared_offset = relative_row * fetch_block_row_size + relative_col_offset;
+        auto const input_src = input_data[absolute_col] + col_size * absolute_row;
+
+        // copy the element from global memory
+        switch (col_size) {
+          case 2:
```

Comment:
I assume this is an optimization to not require `cuda::memcpy_async` to do its own alignment. If that's the case, what kind of perf gains do we get with this change?

Reply:
An excellent question and one I will determine. The general idea is that it is "better" to tell `memcpy_async` the alignment so it can optimize the memory transfer, but I'm not sure the exact benefit here.

Reply:
I tested this and it is about a 2.5% speedup to have this. I'm inclined to keep it.